### DISCUSSION TITLE
parallel cg_user_data_write to different nodes per rank, gopath behavior

### DISCUSSION THREAD

**Discussion by mlohry** [CGNS/CGNS#803] [CGNStalk]
I'm attempting to write arbitrary user data arrays from each rank to a single file output. To write to one global array, the process is e.g.
```
cg_gopath(file_index, "/Base");
cg_user_data_write("Data");
cg_gopath(file_index, "/Base/Data");
```
to create the path `/Base/Data`, and then one calls `cgp_array_write(...)` and `cgp_array_write_data`.

I would like to keep each rank's data stored in separate paths, e.g. `/Base/Partition0/Data` and `/Base/Partition1/Data`. However it seems that `gopath` is globally collective, so I can't have each rank writing to a different array or node. Is there any way to accomplish this?

**Comment by jrwrigh** [comment]
I don't think `cg_gopath` is collective, but `cgp_array_write` and `cgp_array_write_data` are collective. They meant to be called on the same CGNS node.

**Reply by brtnfld** [reply]
In the end, cg_gopath will call H5Gopen in HDF5, which normally has to be a collective call unless the target object will not be modified; then, it can be independently called.

Generally, any CGNS function that creates an "entry" in the CGNS file (e.g., `cgp_array_write`) needs to be collective. HDF5 calls like group and dataset creation APIs must be called collectively. cgp_array_write_data does not have to be collective if independent I/O is used.

We should add a note in the documentation which APIs are collective.

**Comment by brtnfld** [comment]
What scale do you plan to ramp up to? If you plan to create entries for each rank, you might not be happy with the performance once you get into 10-100k ranks. This usually becomes an issue because the metadata/metadata coordination between the ranks will begin to play a role, and parallel filesystems are unhappy when dealing with metadata; they are tailored for large writes. You will have to tune CGNS (which means tuning HDF5) to minimize this at large scales.

Do you know what the CGNS's file "structure" will be ahead of time (i.e., everything is predetermined, minus the raw data)?

**Reply by mlohry** [reply]
Hi Scot, we have needs for two scales/implementations. We're running on frontier and aurora and other big systems where 70k+ ranks and ~100B vertices are expected, and I've implemented file-per-rank output for this where each rank outputs a fully-fledged local mesh+solution+user data arrays. As I understand it from past discussions this is an appropriate approach for this scale, but please let me know otherwise.

On the other end of more routine simulations we'd like to have single file output as an option. In either case a priority is their use as restart files; i.e. we want to write all halo data and solution/mesh arrays in partition local ordering. So with single file pcgns output I can think of two ways to tackle this -- one zone per partition where each zone looks like those from file-per-rank, or write global arrays of all data, and possibly adjust a global connectivity array to provide a globally indexed view (although that would purely be for viz elsewhere). I take it the latter is the suggested approach? I had a request that any rank be able to write arbitrary bytestream arrays hence hitting the original issue.

>Do you know what the CGNS's file "structure" will be ahead of time (i.e., everything is predetermined, minus the raw data)?

I know the tree-like structure I would like to end up with, if that's what you mean, like this for two partitions writing to one CGNS:

```
% h5ls -r ./ex2_mesh_and_solution.cgns 
/                        Group
/\ format                Dataset {15}
/\ hdf5version           Dataset {33}
/Base                    Group
/Base/\ data             Dataset {2}
/Base/Partition_000000   Group
/Base/Partition_000000/\ data Dataset {3, 1}
/Base/Partition_000000/Elem_HEXA_8 Group
/Base/Partition_000000/Elem_HEXA_8/\ data Dataset {2}
/Base/Partition_000000/Elem_HEXA_8/ElementConnectivity Group
/Base/Partition_000000/Elem_HEXA_8/ElementConnectivity/\ data Dataset {8}
/Base/Partition_000000/Elem_HEXA_8/ElementRange Group
/Base/Partition_000000/Elem_HEXA_8/ElementRange/\ data Dataset {2}
/Base/Partition_000000/Elem_PENTA_6 Group
/Base/Partition_000000/Elem_PENTA_6/\ data Dataset {2}
/Base/Partition_000000/Elem_PENTA_6/ElementConnectivity Group
/Base/Partition_000000/Elem_PENTA_6/ElementConnectivity/\ data Dataset {12306}
/Base/Partition_000000/Elem_PENTA_6/ElementRange Group
/Base/Partition_000000/Elem_PENTA_6/ElementRange/\ data Dataset {2}
/Base/Partition_000000/Elem_PYRA_5 Group
/Base/Partition_000000/Elem_PYRA_5/\ data Dataset {2}
/Base/Partition_000000/Elem_PYRA_5/ElementConnectivity Group
/Base/Partition_000000/Elem_PYRA_5/ElementConnectivity/\ data Dataset {65}
/Base/Partition_000000/Elem_PYRA_5/ElementRange Group
/Base/Partition_000000/Elem_PYRA_5/ElementRange/\ data Dataset {2}
/Base/Partition_000000/Elem_TETRA_4 Group
/Base/Partition_000000/Elem_TETRA_4/\ data Dataset {2}
/Base/Partition_000000/Elem_TETRA_4/ElementConnectivity Group
/Base/Partition_000000/Elem_TETRA_4/ElementConnectivity/\ data Dataset {16760}
/Base/Partition_000000/Elem_TETRA_4/ElementRange Group
/Base/Partition_000000/Elem_TETRA_4/ElementRange/\ data Dataset {2}
/Base/Partition_000000/GridCoordinates Group
/Base/Partition_000000/GridCoordinates/CoordinateX Group
/Base/Partition_000000/GridCoordinates/CoordinateX/\ data Dataset {2141}
/Base/Partition_000000/GridCoordinates/CoordinateY Group
/Base/Partition_000000/GridCoordinates/CoordinateY/\ data Dataset {2141}
/Base/Partition_000000/GridCoordinates/CoordinateZ Group
/Base/Partition_000000/GridCoordinates/CoordinateZ/\ data Dataset {2141}
/Base/Partition_000000/ZoneType Group
/Base/Partition_000000/ZoneType/\ data Dataset {12}
/Base/Partition_000001   Group
/Base/Partition_000001/\ data Dataset {3, 1}
/Base/Partition_000001/Elem_HEXA_8 Group
/Base/Partition_000001/Elem_HEXA_8/\ data Dataset {2}
/Base/Partition_000001/Elem_HEXA_8/ElementConnectivity Group
/Base/Partition_000001/Elem_HEXA_8/ElementConnectivity/\ data Dataset {8}
/Base/Partition_000001/Elem_HEXA_8/ElementRange Group
/Base/Partition_000001/Elem_HEXA_8/ElementRange/\ data Dataset {2}
/Base/Partition_000001/Elem_PENTA_6 Group
/Base/Partition_000001/Elem_PENTA_6/\ data Dataset {2}
/Base/Partition_000001/Elem_PENTA_6/ElementConnectivity Group
/Base/Partition_000001/Elem_PENTA_6/ElementConnectivity/\ data Dataset {12306}
/Base/Partition_000001/Elem_PENTA_6/ElementRange Group
/Base/Partition_000001/Elem_PENTA_6/ElementRange/\ data Dataset {2}
/Base/Partition_000001/Elem_PYRA_5 Group
/Base/Partition_000001/Elem_PYRA_5/\ data Dataset {2}
/Base/Partition_000001/Elem_PYRA_5/ElementConnectivity Group
/Base/Partition_000001/Elem_PYRA_5/ElementConnectivity/\ data Dataset {65}
/Base/Partition_000001/Elem_PYRA_5/ElementRange Group
/Base/Partition_000001/Elem_PYRA_5/ElementRange/\ data Dataset {2}
/Base/Partition_000001/Elem_TETRA_4 Group
/Base/Partition_000001/Elem_TETRA_4/\ data Dataset {2}
/Base/Partition_000001/Elem_TETRA_4/ElementConnectivity Group
/Base/Partition_000001/Elem_TETRA_4/ElementConnectivity/\ data Dataset {16760}
/Base/Partition_000001/Elem_TETRA_4/ElementRange Group
/Base/Partition_000001/Elem_TETRA_4/ElementRange/\ data Dataset {2}
/Base/Partition_000001/GridCoordinates Group
/Base/Partition_000001/GridCoordinates/CoordinateX Group
/Base/Partition_000001/GridCoordinates/CoordinateX/\ data Dataset {2141}
/Base/Partition_000001/GridCoordinates/CoordinateY Group
/Base/Partition_000001/GridCoordinates/CoordinateY/\ data Dataset {2141}
/Base/Partition_000001/GridCoordinates/CoordinateZ Group
/Base/Partition_000001/GridCoordinates/CoordinateZ/\ data Dataset {2141}
/Base/Partition_000001/ZoneType Group
/Base/Partition_000001/ZoneType/\ data Dataset {12}
/Base/SerializedData     Group
/Base/SerializedData/Partition_000000 Group
/Base/SerializedData/Partition_000000/Data Group
/Base/SerializedData/Partition_000000/Data/i Group
/Base/SerializedData/Partition_000000/Data/i/\ data Dataset {1}
/Base/SerializedData/Partition_000000/Data/str_data Group
/Base/SerializedData/Partition_000000/Data/str_data/\ data Dataset {11}
/Base/SerializedData/Partition_000000/Data/v Group
/Base/SerializedData/Partition_000000/Data/v/\ data Dataset {10}
/Base/SerializedData/Partition_000001 Group
/Base/SerializedData/Partition_000001/Data Group
/Base/SerializedData/Partition_000001/Data/i Group
/Base/SerializedData/Partition_000001/Data/i/\ data Dataset {1}
/Base/SerializedData/Partition_000001/Data/str_data Group
/Base/SerializedData/Partition_000001/Data/str_data/\ data Dataset {11}
/Base/SerializedData/Partition_000001/Data/v Group
/Base/SerializedData/Partition_000001/Data/v/\ data Dataset {10}
/CGNSLibraryVersion      Group
/CGNSLibraryVersion/\ data Dataset {1}
```
where the `SerializedData` contains any arbitrary array a rank cares to write. I'm all ears as to a better approach. Thanks!

**Reply by brtnfld** [reply]
> Hi Scot, we have needs for two scales/implementations. We're running on frontier and aurora and other big systems where 70k+ ranks and ~100B vertices are expected, and I've implemented file-per-rank output for this where each rank outputs a fully-fledged local mesh+solution+user data arrays. As I understand it from past discussions this is an appropriate approach for this scale, but please let me know otherwise.

If this works for you and provides good performance, and you don't mind the challenge of managing many files, then it seems like a suitable option. I understand the Frontier File System is optimized for this method rather than a single-file approach by default.

**Reply by jrwrigh** [reply]
> We're running on frontier and aurora and other big systems where 70k+ ranks and ~100B vertices are expected, and I've implemented file-per-rank output for this where each rank outputs a fully-fledged local mesh+solution+user data arrays. As I understand it from past discussions this is an appropriate approach for this scale, but please let me know otherwise.

HONEE/PETSc is getting really good scaling for single file parallel on Aurora, just using LFS (not DAOS yet). We've only run to 36k ranks and 18B nodes (90B total solution scalars), but the read and write only takes a dozen or so seconds.

**Reply by jrwrigh** [reply]
I see from your GitHub profile PETSc is mentioned. IDK if you're using the DMPlex interface, but we're using the parallel reading and writing I helped implement over there. It's limited to monotopology, uniform function space (i.e. no velocity-discontinuous pressure pairs), and FEM at the moment, but that's more because I don't have a use case (and importantly a real-world verification test) than a limitation of the implementation.

**Reply by brtnfld** [reply]
> On the other end of more routine simulations we'd like to have single file output as an option. In either case a priority is their use as restart files; i.e. we want to write all halo data and solution/mesh arrays in partition local ordering. So with single file pcgns output I can think of two ways to tackle this -- one zone per partition where each zone looks like those from file-per-rank, or write global arrays of all data, and possibly adjust a global connectivity array to provide a globally indexed view (although that would purely be for viz elsewhere). I take it the latter is the suggested approach? I had a request that any rank be able to write arbitrary bytestream arrays hence hitting the original issue.



The second approach, which is more commonly used, is preferred because it avoids introducing parallel elements into the CGNS file. In contrast, the first approach requires handling the case of a different number of ranks reading the file, making it less straightforward.

Regarding performance, I have yet to compare the method of having all ranks contribute to a section of a single zone versus breaking the zone into multiple zones per rank and writing it that way. In the zone-per-rank case, you face limitations since each rank must perform independent writing, sometimes resulting in poorer performance than collective writing. A one-to-one comparison is necessary for a comprehensive evaluation.

I have modified the `benchmark_hdf5.c` code in the source to align with your suggestion. I create the CGNS file structure on one rank and then close the file. This method has demonstrated good scalability as the number of ranks increases, although we usually only observe a few seconds difference. After closing the file, I opened it on all ranks, allowing each rank to write the raw data to its respective zone.

[benchmark_part.tgz](https://github.com/user-attachments/files/18049193/benchmark_part.tgz)

Consider utilizing the HDF5 subfiling VFD in CGNS if you use the one-zone method. On Frontier, this approach allows you to leverage node-local storage, which is reasonably fast and scales nearly linearly with the number of nodes. At larger scales, you can expect speeds up to TiB/s. 

You can then transfer the node-local subfiles to the parallel file system, consolidating them into a single CGNS file. I am still experimenting with making this transfer more efficient and better justify its use over simply writing to a single shared file on the parallel file system.

I/O is a complex subject; we haven't even touched on DAOS 😬.

**Reply by mlohry** [reply]
> I see from your GitHub profile PETSc is mentioned. IDK if you're using the DMPlex interface, but we're using the parallel reading and writing I helped implement over there.

No PETSc being used for this, but good to know, thanks!

**Reply by smguzik** [reply]
In the benchmark_part example, why do you switch to independent I/O if every rank is participating in the parallel writes?

**Reply by brtnfld** [reply]
Since each rank is opening and writing to a separate HDF5 dataset, it has to be done independently. Collective would require all the ranks to call the write command for each dataset being written to, as HDF5 expects all the ranks in the MPI communicator to perform in the dataset write for collective operations.

**Reply by smguzik** [reply]
There are a couple of issues with the library then:

1. The documentation is confusing and not sufficiently descriptive.  For cgp_pio_mode, it currently states,

> Sets the mode for parallel data reads and writes. The default value is CGP_COLLECTIVE, which allows any number of processes to access the data. When set to CGP_COLLECTIVE, all processes must access the data.

That seems to say two different things.  Do you agree with the wording of the following instead,

> Sets the mode for parallel data reads and writes.  The default value is CGP_COLLECTIVE, which may allow for higher performance parallel reads and writes.  However, each process must participate in a read/write to the same HDF5 dataset.  This means that the each process must call the cgp* functions and request the same node (as specified by file, base, zone, etc. index numbers or using one of the cg_goto variants).  An alternative option is CGP_INDEPENDENT, which allows any number of processes to read/write to any location in the file.

2. The `*paroverzone*` examples in src/Test_UserGuideCode/C_code_parallel are incorrect and should set pio mode to CGP_INDEPENDENT.

3. Those examples currently work using multiple processes on a shared memory system writing to a serial file system.  I am not sure why, but I'm guessing that errors will appear when switching to distributed memory and/or a parallel file system.  If possible, I think it would be better to return an error when the mode is used incorrectly.  In the current setup, someone might develop on a workstation, be fooled into thinking that the code is correct, transfer to a supercomputer, and encounter unexpected errors.  Is there a way we can detect/force this error short of separately checking each read/write node path?

4. If the checking is possible, can we have an auto mode that uses collective I/O when accessing the same dataset, and otherwise automatically switches to independent I/O?

**Comment by smguzik** [comment]
> I create the CGNS file structure on one rank and then close the file. This method has demonstrated good scalability as the number of ranks increases, although we usually only observe a few seconds difference. After closing the file, I opened it on all ranks, allowing each rank to write the raw data to its respective zone.

Scot, just for the sake of discussion, if one did want to write the zones in parallel, does HDF5 support that?  Or would we have to create an array-like structure of zone data (and probably have to introduce a new type into the SIDS).  On the read side, it is annoying that the library reads every zone's meta-data upon opening a file when it is probable that a rank may only need to access a few.

**Reply by brtnfld** [reply]
> Scot, just for the sake of discussion, if one did want to write the zones in parallel, does HDF5 support that? Or would we have to create an array-like structure of zone data (and probably have to introduce a new type into the SIDS). On the read side, it is annoying that the library reads every zone's meta-data upon opening a file when it is probable that a rank may only need to access a few.

cg_zone_write creates datasets, requiring all ranks to call it for each zone created. Essentially, you need to use the if(comm_rank == 0) contents, and that content called by all ranks. 

It would be reasonable to add an option to cgp_open that only opens the cgns file, but I'm not sure about the scope of the work necessary to open and read the information of a specific zone.
